File: FSUIKit/Classes/FSViewManager.m
//
//  FSViewManager.m
//  ShareEconomy
//
//  Created by FudonFuchina on 16/4/13.
//  Copyright © 2016年 FudonFuchina. All rights reserved.
//

#import "FSViewManager.h"

@implementation FSViewManager

+ (UIView *)viewWithFrame:(CGRect)frame backColor:(UIColor *)color
{
    UIView *view = [[UIView alloc] initWithFrame:frame];
    if (color) {
        view.backgroundColor = color;
    }
    return view;
}

+ (UIView *)seprateViewWithFrame:(CGRect)frame {
    UIView *view = [[UIView alloc] initWithFrame:frame];
    CGFloat rgb = 230 / 255.0;
    view.backgroundColor = [UIColor colorWithRed:rgb green:rgb blue:rgb alpha:1];
    return view;
}

+ (UIBarButtonItem *)bbiWithTitle:(NSString *)title target:(id)target action:(SEL)selector
{
    UIBarButtonItem *bbi = [[UIBarButtonItem alloc] initWithTitle:title style:UIBarButtonItemStylePlain target:target action:selector];
    return bbi;
}

+ (UIBarButtonItem *)bbiWithSystemType:(UIBarButtonSystemItem)type target:(id)target action:(SEL)selector
{
    UIBarButtonItem *bbi = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:type target:target action:selector];
    return bbi;
}

+ (UISegmentedControl *)segmentedControlWithTitles:(NSArray<NSString *> *)titles target:(id)target action:(SEL)selector
{
    UISegmentedControl *control = [[UISegmentedControl alloc] initWithItems:titles];
    [control addTarget:target action:selector forControlEvents:UIControlEventValueChanged];
    return control;
}

+ (UIButton *)buttonWithFrame:(CGRect)frame title:(NSString *)title titleColor:(UIColor *)color backColor:(UIColor *)backColor font:(UIFont *)font tag:(NSInteger)tag target:(id)target selector:(SEL)selector{
    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    button.frame = frame;
    if (title) {
        [button setTitle:title forState:UIControlStateNormal];
    }
    if (color) {
        [button setTitleColor:color forState:UIControlStateNormal];
    }else{
        [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    }
    if (backColor) {
        button.backgroundColor = backColor;
    }
    if (target && selector) {
        [button addTarget:target action:selector forControlEvents:UIControlEventTouchUpInside];
    }
    if (font) {
        button.titleLabel.font =  font;
    }
    if (tag) {
        button.tag = tag;
    }
    button.layer.cornerRadius = 3;
    return button;
}

+ (UIButton *)submitButtonWithTop:(CGFloat)top tag:(NSInteger)tag target:(id)target selector:(SEL)selector {
    UIColor *color = [UIColor colorWithRed:18/255.0 green:152/255.0 blue:233/255.0 alpha:1];
    UIButton *button = [self buttonWithFrame:CGRectMake(20, top, UIScreen.mainScreen.bounds.size.width - 40, 44) title:@"提交" titleColor:[UIColor whiteColor] backColor:color font:nil tag:tag target:target selector:selector];
    button.layer.cornerRadius = 3;
    return button;
}

+ (UIBarButtonItem *)barButtonItemWithCustomButton:(UIButton *)button
{
    UIBarButtonItem *bbi = [[UIBarButtonItem alloc] initWithCustomView:button];
    return bbi;
}

+ (UIBarButtonItem *)barButtonItemWithTitle:(NSString *)title target:(id)target selector:(SEL)selector tintColor:(UIColor *)tintColor
{
    UIBarButtonItem *bbi = [[UIBarButtonItem alloc] initWithTitle:title style:UIBarButtonItemStylePlain target:target action:selector];
    if (tintColor) {
        bbi.tintColor = tintColor;
    }
    return bbi;
}

+ (UITableView *)tableViewWithFrame:(CGRect)frame delegate:(id)delegate style:(UITableViewStyle)style footerView:(UIView *)footerView
{
    UITableView *tableView = [[UITableView alloc] initWithFrame:frame style:style];
    tableView.delegate = delegate;
    tableView.dataSource = delegate;
    if (footerView) {
        tableView.tableFooterView = footerView;
    }
    return tableView;
}

+ (UILabel *)labelWithFrame:(CGRect)frame
                       text:(NSString *)text
                  textColor:(UIColor *)textColor
                  backColor:(UIColor *)backColor
                       font:(UIFont *)font
              textAlignment:(NSTextAlignment)textAlignment
{
    UILabel *label = [[UILabel alloc] initWithFrame:frame];
    if (text) {
        label.text = text;
    }
    if (textColor) {
        label.textColor = textColor;
    }
    if (backColor) {
        label.backgroundColor = backColor;
    }
    if (font) {
        label.font = font;
    }
    label.textAlignment = textAlignment;
    return label;
}

+ (UILabel *)suojinLabelWithSpace:(CGFloat)space frame:(CGRect)rect textColor:(UIColor *)textColor text:(NSString *)text
{
    UILabel *label = [[UILabel alloc] initWithFrame:rect];
    
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:text?text:@""];
    if ([textColor isKindOfClass:UIColor.class]) {
        [attributedString addAttribute:NSForegroundColorAttributeName value:textColor range:NSMakeRange(0, attributedString.length)];
    }
    
    NSMutableParagraphStyle *style = [[NSMutableParagraphStyle alloc] init];
    //    style.headIndent = MAX(space, 0);//整体缩进
    style.firstLineHeadIndent = space;  // 首行缩进
    //    style.lineSpacing = 10;//行距
    [attributedString addAttribute:NSParagraphStyleAttributeName value:style range:NSMakeRange(0, attributedString.length)];
    label.attributedText = attributedString;
    return label;
}

+ (FSTapLabel *)tapLabelWithFrame:(CGRect)frame
                          text:(NSString *)text
                     textColor:(UIColor *)textColor
                     backColor:(UIColor *)backColor
                          font:(UIFont *)font
                 textAlignment:(NSTextAlignment)textAlignment
                         block:(FSTapLabelBlock)block
{
    FSTapLabel *label = [[FSTapLabel alloc] initWithFrame:frame];
    if (text) {
        label.text = text;
    }
    if (textColor) {
        label.textColor = textColor;
    }
    if (backColor) {
        label.backgroundColor = backColor;
    }
    if (font) {
        label.font = font;
    }
    if (block) {
        label.block = block;
    }
    label.textAlignment = textAlignment;
    return label;
}

+ (FSTapCell *)tapCellWithText:(NSString *)text
                      textColor:(UIColor *)textColor
                          font:(UIFont *)font
                    detailText:(NSString *)detailText
                   detailColor:(UIColor *)detailColor
                    detailFont:(UIFont *)detailFont
                          block:(TapCellBlock)block {
    return [self tapCellWithText: text textColor: textColor font: font detailText: detailText detailColor: detailColor detailFont: detailFont frame: CGRectMake(0, 0, UIScreen.mainScreen.bounds.size.width, 44) block: block];
}

+ (FSTapCell *)tapCellWithText:(NSString *)text
                      textColor:(UIColor *)textColor
                          font:(UIFont *)font
                    detailText:(NSString *)detailText
                   detailColor:(UIColor *)detailColor
                    detailFont:(UIFont *)detailFont
                         frame:(CGRect)frame
                          block:(TapCellBlock)block {
    FSTapCell *cell = [[FSTapCell alloc] initWithFrame: frame];
    if (text) {
        cell.textLabel.text = text;
    }
    if (textColor) {
        cell.textLabel.textColor = textColor;
    }
    if (font) {
        cell.textLabel.font = font;
    }
    if (detailText) {
        cell.detailTextLabel.text = detailText;
    }
    if (detailColor) {
        cell.detailTextLabel.textColor = detailColor;
    }
    if (detailFont) {
        cell.detailTextLabel.font = detailFont;
    }
    if (block) {
        cell.click = block;
    }
    cell.frame = CGRectMake(cell.frame.origin.x, cell.frame.origin.y, UIScreen.mainScreen.bounds.size.width, cell.frame.size.height);
    return cell;
}

+ (UITextField *)textFieldWithFrame:(CGRect)frame placeholder:(NSString *)holder textColor:(UIColor *)textColor backColor:(UIColor *)backColor {
    UITextField *textField = [[UITextField alloc] initWithFrame:frame];
    if (holder) {
        textField.placeholder = holder;
    }
    if (textColor) {
        textField.textColor = textColor;
    }
    if (backColor) {
        textField.backgroundColor = backColor;
    }
    textField.autocorrectionType = UITextAutocorrectionTypeNo;
    textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    textField.autocapitalizationType = UITextAutocapitalizationTypeNone;
    return textField;
}

+ (UITextField *)textFieldWithFrame:(CGRect)frame placeholder:(NSString *)holder textColor:(UIColor *)textColor onlyChars:(BOOL)only
{
    UITextField *textField = [[UITextField alloc] initWithFrame:frame];
    if (holder) {
        textField.placeholder = holder;
    }
    if (textColor) {
        textField.textColor = textColor;
    }
    textField.autocorrectionType = UITextAutocorrectionTypeNo;
    textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    textField.autocapitalizationType = UITextAutocapitalizationTypeNone;
    if (only) {
        textField.keyboardType = UIKeyboardTypeASCIICapable;
    }
    return textField;
}

+ (UIImageView *)imageViewWithFrame:(CGRect)frame imageName:(NSString *)imageName
{
    UIImageView *imageView = [[UIImageView alloc] initWithFrame:frame];
    if (imageName) {
        UIImage *image = [UIImage imageNamed:imageName];
        if (image) {
            imageView.image = image;
        }
    }
    imageView.contentMode = UIViewContentModeScaleAspectFit;
    return imageView;
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSViewManager.swift
// FSViewManagerS.swift
// Translated from FSViewManager.h/m

import UIKit

public let FS_LineThickness: CGFloat = 0.5

@objcMembers
open class FSViewManager: UIView {
    
    public static func view(withFrame frame: CGRect, backColor color: UIColor?) -> UIView {
        let view = UIView(frame: frame)
        if let c = color {
            view.backgroundColor = c
        }
        return view
    }
    
    public static func imageView(withFrame frame: CGRect, imageName: String?) -> UIImageView {
        let imageView = UIImageView(frame: frame)
        if let name = imageName, let image = UIImage(named: name) {
            imageView.image = image
        }
        imageView.contentMode = .scaleAspectFit
        return imageView
    }
    
    public static func separateView(withFrame frame: CGRect) -> UIView {
        let view = UIView(frame: frame)
        let rgb: CGFloat = 230 / 255.0
        view.backgroundColor = UIColor(red: rgb, green: rgb, blue: rgb, alpha: 1)
        return view
    }
    
    public static func bbi(withTitle title: String, target: Any?, action selector: Selector) -> UIBarButtonItem {
        return UIBarButtonItem(title: title, style: .plain, target: target, action: selector)
    }
    
    public static func bbi(withSystemType type: UIBarButtonItem.SystemItem, target: Any?, action selector: Selector) -> UIBarButtonItem {
        return UIBarButtonItem(barButtonSystemItem: type, target: target, action: selector)
    }
    
    public static func segmentedControl(withTitles titles: [String], target: Any?, action selector: Selector) -> UISegmentedControl {
        let control = UISegmentedControl(items: titles)
        control.addTarget(target, action: selector, for: .valueChanged)
        return control
    }
    
    public static func button(withFrame frame: CGRect, title: String?, titleColor color: UIColor?, backColor: UIColor?, font: UIFont?, tag: Int, target: Any?, selector: Selector?) -> UIButton {
        let button = UIButton(type: .system)
        button.frame = frame
        if let t = title {
            button.setTitle(t, for: .normal)
        }
        button.setTitleColor(color ?? .white, for: .normal)
        if let bc = backColor {
            button.backgroundColor = bc
        }
        if let t = target, let s = selector {
            button.addTarget(t, action: s, for: .touchUpInside)
        }
        if let f = font {
            button.titleLabel?.font = f
        }
        button.tag = tag
        button.layer.cornerRadius = 3
        return button
    }
    
    public static func submitButton(withTop top: CGFloat, tag: Int, target: Any?, selector: Selector) -> UIButton {
        let color = UIColor(red: 18/255.0, green: 152/255.0, blue: 233/255.0, alpha: 1)
        let button = self.button(withFrame: CGRect(x: 20, y: top, width: UIScreen.main.bounds.width - 40, height: 44), title: "提交", titleColor: .white, backColor: color, font: nil, tag: tag, target: target, selector: selector)
        button.layer.cornerRadius = 3
        return button
    }
    
    public static func barButtonItem(withCustomButton button: UIButton) -> UIBarButtonItem {
        return UIBarButtonItem(customView: button)
    }
    
    public static func barButtonItem(withTitle title: String, target: Any?, selector: Selector, tintColor: UIColor?) -> UIBarButtonItem {
        let bbi = UIBarButtonItem(title: title, style: .plain, target: target, action: selector)
        if let tc = tintColor {
            bbi.tintColor = tc
        }
        return bbi
    }
    
    public static func tableView(withFrame frame: CGRect, delegate: (UITableViewDelegate & UITableViewDataSource)?, style: UITableView.Style, footerView: UIView?) -> UITableView {
        let tableView = UITableView(frame: frame, style: style)
        tableView.delegate = delegate
        tableView.dataSource = delegate
        if let fv = footerView {
            tableView.tableFooterView = fv
        }
        return tableView
    }
    
    @available(iOS 6.0, *)
    public static func label(withFrame frame: CGRect, text: String?, textColor: UIColor?, backColor: UIColor?, font: UIFont?, textAlignment: NSTextAlignment) -> UILabel {
        let label = UILabel(frame: frame)
        label.text = text
        if let tc = textColor {
            label.textColor = tc
        }
        if let bc = backColor {
            label.backgroundColor = bc
        }
        if let f = font {
            label.font = f
        }
        label.textAlignment = textAlignment
        return label
    }
    
    public static func suojinLabel(withSpace space: CGFloat, frame rect: CGRect, textColor: UIColor?, text: String?) -> UILabel {
        let label = UILabel(frame: rect)
        
        let attributedString = NSMutableAttributedString(string: text ?? "")
        if let tc = textColor {
            attributedString.addAttribute(.foregroundColor, value: tc, range: NSRange(location: 0, length: attributedString.length))
        }
        
        let style = NSMutableParagraphStyle()
        style.firstLineHeadIndent = space
        attributedString.addAttribute(.paragraphStyle, value: style, range: NSRange(location: 0, length: attributedString.length))
        label.attributedText = attributedString
        return label
    }
    
    @available(iOS 6.0, *)
    public static func tapLabel(withFrame frame: CGRect, text: String?, textColor: UIColor?, backColor: UIColor?, font: UIFont?, textAlignment: NSTextAlignment, block: FSTapLabelBlockS?) -> FSTapLabel {
        let label = FSTapLabel(frame: frame)
        label.text = text
        if let tc = textColor {
            label.textColor = tc
        }
        if let bc = backColor {
            label.backgroundColor = bc
        }
        if let f = font {
            label.font = f
        }
        label.block = block
        label.textAlignment = textAlignment
        return label
    }
    
    public static func tapCell(withText text: String?, textColor: UIColor?, font: UIFont?, detailText: String?, detailColor: UIColor?, detailFont: UIFont?, block: TapCellBlock?) -> FSTapCell {
        return tapCell(withText: text, textColor: textColor, font: font, detailText: detailText, detailColor: detailColor, detailFont: detailFont, frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 44), block: block)
    }
    
    public static func tapCell(withText text: String?, textColor: UIColor?, font: UIFont?, detailText: String?, detailColor: UIColor?, detailFont: UIFont?, frame: CGRect, block: TapCellBlock?) -> FSTapCell {
        let cell = FSTapCell(frame: frame)
        if let t = text {
            cell.textLabel.text = t
        }
        if let tc = textColor {
            cell.textLabel.textColor = tc
        }
        if let f = font {
            cell.textLabel.font = f
        }
        if let dt = detailText {
            cell.detailTextLabel.text = dt
        }
        if let dc = detailColor {
            cell.detailTextLabel.textColor = dc
        }
        if let df = detailFont {
            cell.detailTextLabel.font = df
        }
        cell.click = block
        return cell
    }
    
    public static func textField(withFrame frame: CGRect, placeholder holder: String?, textColor: UIColor?, backColor: UIColor?) -> UITextField {
        let textField = UITextField(frame: frame)
        textField.placeholder = holder
        if let tc = textColor {
            textField.textColor = tc
        }
        if let bc = backColor {
            textField.backgroundColor = bc
        }
        textField.autocorrectionType = .no
        textField.clearButtonMode = .whileEditing
        textField.autocapitalizationType = .none
        return textField
    }
    
    public static func textField(withFrame frame: CGRect, placeholder holder: String?, textColor: UIColor?, onlyChars only: Bool) -> UITextField {
        let textField = UITextField(frame: frame)
        textField.placeholder = holder
        if let tc = textColor {
            textField.textColor = tc
        }
        textField.autocorrectionType = .no
        textField.clearButtonMode = .whileEditing
        textField.autocapitalizationType = .none
        if only {
            textField.keyboardType = .asciiCapable
        }
        return textField
    }
}
